### ISSUE TITLE
Bad performance when dragging with a mouse

### ISSUE THREAD

**Issue by isfalcantara** [aseprite/aseprite#2065]
For some reason dragging Aseprite's internal windows has a terrible performance when using a mouse. When using my laptop's touchpad, dragging a window becomes way faster and more responsive.

[Here](https://youtu.be/p-qXV801hmE) is a video showing the bug in action.

- Aseprite version: v1.2.10-x64, Steam
- System: Linux, kernel 5.1.0-1-MANJARO, Manjaro Linux

**Comment by 099waldo** [comment]
Same thing is happening for me. 

Most recent dev version. Ubuntu 18.04

**Comment by dacap** [comment]
@isfalcantara does your mouse is a high DPI? (some specific model?) or just a regular mouse?

@099waldo do you experience the same different between mouse vs trackpad? or both are slow?

**Comment by 099waldo** [comment]
Yeah for me it looks almost identical to @isfalcantara 's video. The trackpad is faster but still laggy and the mouse is much more laggy than the trackpad. I've tested on multiple machines but they all yield the same results.

**Comment by isfalcantara** [comment]
The mouse used in the video was a *Redragon Phoenix Chroma M702-2*, with 10000DPI. I've also tested with a *Logitech M170* wireless mouse, which I believe has around 1000DPI, with very similar results. Today I tried to observe the issue in another computer, and the lag is very small compared to what I previously experienced. The old computer has the following specs:

- **CPU**: Intel core i5 6200U
- **GPU**: Intel HD Graphics 520
- **RAM**: 8GB

and the new machine has the following specs:


- **CPU**: AMD A10 9600P
- **GPU**: AMD R8 M445DX
- **RAM**: 4GB

Unfortunately I no longer have access to my previous computer, so I can not conduct further tests.

**Comment by flamendless** [comment]
It happens to me as well.

Manjaro Linux.

**Comment by hollunder** [comment]
For me it is much worse, the window is chugging after the mouse and often moving into the opposite direction.

This is from aseprite-git 1.2.12.r1.g86ffa37fd on up-to-date Arch Linux using this build script: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=aseprite-git
Sidenote: window capture in OBS didn't work properly for this version.
Video: https://tube.tchncs.de/videos/watch/b7142816-78ef-4e70-a897-b9294b1d51f5

It is somewhat better if I run it through Steam (Runtime and Native version) (V1.2.12-x64) where the window just tears badly while being dragged. This was recorded with 60 FPS but the tearing isn't really visible on video. Maybe it's driver dependent? AMDGPU in my case.
Video: https://tube.tchncs.de/videos/watch/d09dbe4e-8a6f-4d9e-ad2f-993b12997779

The problem has persisted since I know Aseprite, so the last couple years.

**Comment by saranvdev** [comment]
Related to #2089 and #2098. This is basically the same bug. Whether moving internal floating windows or moving a selection. Doesn't matter.

As long as we move it with the mouse, lags like hell.
But with the touchpad it works correctly.

Fix please, it can't be that hard to just redefine mouse to act like touchpad or something.

**Comment by saranvdev** [comment]
UPDATE. Found a fix.

Enable "Cursors->Use native mouse cursors"  in preferences. The lag disappears.

**Comment by dacap** [comment]
@saranvdev what specific version/commit of Aseprite where you using? is it still compiled with the Allegro back-end? Because that might be a problem.

**Comment by saranvdev** [comment]
The newest stable 1.2.21, Manjaro Linux
The lag seems to be related with linux and that option mentioned above.

**Comment by dacap** [comment]
It's good to know a fix. It's strange because the non-native cursors use the Xcursor library, which changes the native cursor to other bitmaps, but maybe something is wrong or not as good as it could be (maybe non-bitmap cursors make everything slowers in some X11 or Wayland implementations, not sure).

**Comment by henrikkudesu** [comment]
> UPDATE. Found a fix.
> 
> Enable "Cursors->Use native mouse cursors" in preferences. The lag disappears.

Thank you so much. Such a silly and old problem and still in 2022 has not been solved :(

**Comment by dacap** [comment]
Some performance fixes related to the default cursor (on sprite) where added since v1.2.28 (Jul 2021) but I'm not sure if this solve some performance issues on Linux (Linux is the most problematic platform because it depends on each distribution):

* https://www.aseprite.org/release-notes/12/#aseprite-v1-2-28
* https://github.com/aseprite/aseprite/issues/2713#issuecomment-877187329
* https://github.com/aseprite/laf/commit/80ec051ecf4b702d769d4b2483e1a34b52368bde

> Such a silly and old problem and still in 2022 has not been solved :(

It would be great to know what version of Aseprite are you using, what device (mouse, touchpad, stylus, etc. each device has its own DPI) and what exact Linux distribution are you using.

Probably in a future we'll just enable the native cursors on Linux because the black&white cursor on sprite looks like an impossible UX to bring correctly on all Linux distribution.

**Comment by hollunder** [comment]
For me these problems seem to be gone meanwhile, but I have yet to determine whether it is due to Aseprite changes or changes on my system.

#### System
I now run Sway (Wayland), which is similar to i3 (X11), which I had used before, but obviously rather different in terms of underlying tech. Aseprite seems to be running through XWayland, so the X11 compatibility layer, rather than natively on Wayland. I will need to see what the behavior is back on i3.

I still use AMDGPU, but it may be that I have a faster GPU (Navi 10) than I used to have (Polaris). I'll need to look into when I got that. I hope the difference doesn't come down to that.

#### Aseprite
1.2.30-3 built through the AUR script: https://aur.archlinux.org/packages/aseprite
1.2.30 as provided through Steam.
Both show the same behavior, described below.
I have yet to try the beta again, last I tried it showed too many regressions to be seriously usable.

#### Behavior
I remember having a lot of weird behavior with the 1:1 window, so this is what I mainly tried today.
1. I can move the window about without disturbing artifacts or lag.
2. I can resize the window without artifacts or lag.
I also tried moving and resizing the Aseprite main window.
3. I can move it about without problems.
4. When resizing the main window it flashes black quite a bit, but as soon as the size has settled it shows the contents just fine. This one I can live with as this is not a common operation for me.

**Comment by hollunder** [comment]
I just tested again, this time on i3/X11. 
Both 1.30 built via AUR and 1.31 from Steam behaved the same.
Same operations as before. What was of note is that
1. Moving the 1:1 window showed noticeable tearing artifacts on that Window
2. Was OK.
3. Was OK.
4. Resizing the main window behaved very similar, but it took longer until after the black rendering the window contents showed again.

What that tells me is that the performance is notably worse on X11 compared to Wayland.
I suspect that it also depends on the GPU how noticeable those issues are.
